python/valis/db/queries.py

```diff
@@ -252,22 +254,29 @@ def carton_program_search(name: str, name_type: str) -> peewee.ModelSelect:
         Either the carton name or the program name
     name_type: str
         Which type you are searching on, either 'carton' or 'program'
+    query : ModelSelect
+        An initial query to extend. If ``None``, a new query with all the unique
+        ``sdss_id``s is created.
 
     Returns
     -------
     peewee.ModelSelect
         the ORM query
     """
-    model = vizdb.SDSSidFlat.select(peewee.fn.DISTINCT(vizdb.SDSSidFlat.sdss_id))\
-                            .join(targetdb.Target,
-                                  on=(targetdb.Target.catalogid == vizdb.SDSSidFlat.catalogid))\
-                            .join(targetdb.CartonToTarget)\
-                            .join(targetdb.Carton)\
-                            .where(getattr(targetdb.Carton, name_type) == name)
-    return vizdb.SDSSidStacked.select().join(
-        model, on=(model.c.sdss_id == vizdb.SDSSidStacked.sdss_id)
-    )
 
+    if query is None:
+        query = vizdb.SDSSidFlat.select(peewee.fn.DISTINCT(vizdb.SDSSidFlat.sdss_id))
+
+    query = (query.join(
+                vizdb.SDSSidFlat,
+                on=(vizdb.SDSSidFlat.sdss_id == vizdb.SDSSidStacked.sdss_id))
+             .join(targetdb.Target,
+                   on=(targetdb.Target.catalogid == vizdb.SDSSidFlat.catalogid))
+             .join(targetdb.CartonToTarget)
+             .join(targetdb.Carton)
+             .where(getattr(targetdb.Carton, name_type) == name))
```

Comment:
We'll likely need to do this for many of the query functions we have, as we add them to the main search.  Is this how you'd recommend modifying them?  Did you try the `select_extend` method?  If so, how did that compare?  

Relatedly, should we be writing our queries differently to make this kind of single-use or dynamic extension easier?

Reply:
The `select_extend()` will only help if we want to return more columns than in the initial query `select`. Do you want this function to also add returning the `program` and `carton` columns? As it is written right now that function can be called with any query that initiates with a `SDSSidStacked` model and it will restrict it to that carton or program.

The problem with the original query was that it would do a subquery to return *all* the unique sdss_ids and then subset to only those in the program or carton. That's a very expensive query and I think not necessary.

Reply:
Eventually I'd like to add the ability in the main search to specify additional columns to return.  These columns in principle could come from any table, which might make things more complicated, but that can be addressed later.

RIght now only a single carton or program can be selected.  It probably doesn't make since to return the carton or program name in that case.  I'd like to eventually move to a multi-select option for program/carton, in which case it would be nice to have those values returned.  I think the query would have to be re-written anyways.

Reply:
Yes, for multiple cartons this query would need to be rewritten. It's fairly easy to change the `.where(getattr(targetdb.Carton, name_type) == name))` to `.where(getattr(targetdb.Carton, name_type).in_(name)))` but I'd wait until that functionality is implemented in the API/Zora since the `IN` statement is less efficient than `==`.

But I did test adding `.select_extend(targetdb.Carton.carton)` for the carton and program after `carton_program_search` has been called and that works fine.